src/dune_rules/install_rules.ml

```diff
@@ -1001,12 +984,45 @@ let scheme_per_ctx_memo =
       let* schemes = Memo.Build.sequential_map packages ~f:(scheme sctx) in
       Scheme.evaluate ~union:Rules.Dir_rules.union (Scheme.all schemes))
 
-let gen_rules sctx ~dir =
-  let* rules, subdirs =
+let symlink_rules sctx ~dir =
+  let+ rules, subdirs =
     let* scheme = Memo.exec scheme_per_ctx_memo sctx in
     Scheme.Evaluated.get_rules scheme ~dir
   in
-  let+ () =
-    Rules.produce_dir ~dir (Option.value ~default:Rules.Dir_rules.empty rules)
+  ( Subdir_set.These subdirs
+  , match rules with
+    | None -> Rules.empty
+    | Some rules -> Rules.of_dir_rules ~dir rules )
+
+let gen_install_alias sctx (package : Package.t) =
+  let ctx = Super_context.context sctx in
+  let name = Package.name package in
+  if not ctx.implicit then
+    let install_fn =
+      Utils.install_file ~package:name ~findlib_toolchain:ctx.findlib_toolchain
+    in
+    let path = Package_paths.build_dir ctx package in
+    let install_alias = Alias.install ~dir:path in
+    let install_file = Path.relative (Path.build path) install_fn in
+    Rules.Produce.Alias.add_deps install_alias
+      (Action_builder.path install_file)
+  else
+    Memo.Build.return ()
+
+let gen_project_rules sctx project =
+  let* () = meta_and_dune_package_rules sctx project in
+  let* only_packages = Only_packages.get () in
+  let packages = Dune_project.packages project in
+  let packages =
+    match only_packages with
+    | None -> packages
+    | Some mask ->
+      Package.Name.Map.merge packages mask ~f:(fun _ p mask ->
+          match (p, mask) with
+          | Some _, Some _ -> p
+          | _ -> None)
   in
-  Subdir_set.These subdirs
+  Memo.Build.sequential_iter (Package.Name.Map.values packages)
```

Comment:
We do have map traversals now.

Reply:
Sounds good, I changed it to a map traversal